### PR TITLE
HTML5 app support, and added utilities for processing HTML, plus some fixes

### PR DESCRIPTION
**Pull request by jamalex**
Note: this PR supersedes (merges in from) #28 and #20, which in turn supersede #19.

Updated #28 with some work I had been doing on HTML5 apps, including utilities and a "predictable zip" function that always produces a zip with the same MD5, as long as the input files are the same.

### CODE REVIEW

**Comment by jamalex** [comment]
Most of the changes here are from the other PRs, but this branch served as a "merge branch" (resolved a number of merge conflicts, of @jayoshih conflicting with her alternate personalities :stuck_out_tongue_winking_eye:), so I'll flag the places that need specific review here.

File: ricecooker/classes/nodes.py [hunk 1]
```diff
@@ -198,7 +202,7 @@ def to_dict(self):
         """
         return {
             "title": self.title,
-            "description": self.description,
+            "description": self.description[:400],
```
Comment (jayoshih) [inline ricecooker/classes/nodes.py]: Can get rid of limit on description length

File: ricecooker/classes/nodes.py [hunk 2]
```diff
@@ -135,9 +139,8 @@ def to_dict(self):
         return {
             "id": self.id.hex,
             "name": self.title,
-            "has_changed": True,
             "thumbnail": self.thumbnail,
-            "description": self.description if self.description is not None else "",
+            "description": self.description[:400] if self.description is not None else "",
```
Comment (jayoshih) [inline ricecooker/classes/nodes.py]: Can get rid of limit on description length

File: ricecooker/classes/nodes.py [hunk 3]
```diff
@@ -454,7 +458,11 @@ def __init__(self, id, title, files, author="", description="", license=None, ex
         files = [] if files is None else files
 
         # Set mastery model defaults if none provided
-        exercise_data = {'mastery_model': exercises.M_OF_N, 'randomize': True, 'm': 3, 'n': 5} if exercise_data is None else exercise_data
+        exercise_data = {} if exercise_data is None else exercise_data
```
Comment (jayoshih) [inline ricecooker/classes/nodes.py]: Enforcing mastery_model on exercise nodes: https://trello.com/c/YJnX5kB4/244-ricecooker-does-not-enforce-mastery-model-on-exercises-but-export-process-does-so-magogenie-channel-cannot-be-exported

File: ricecooker/classes/questions.py [hunk 4]
```diff
@@ -260,6 +292,10 @@ def __init__(self, id, question, correct_answers, all_answers, hints=None):
         set_all_answers = set(all_answers)
         all_answers += [answer for answer in correct_answers if answer not in set_all_answers]
         answers = [self.create_answer(answer, answer in correct_answers) for answer in all_answers]
+        if len(answers) == 0:
```
Comment (jayoshih) [inline ricecooker/classes/questions.py]: More graceful way to handle questions without answers

File: ricecooker/managers.py [hunk 5]
```diff
@@ -385,115 +383,198 @@ def process_tree(self, node, parent=None):
...
+            for f in node[1].files:
+                # Attempt to upload file
+                with  open(config.get_storage_path(f['filename']), 'rb') as file_obj:
+                    response = requests.post(config.file_upload_url(), headers={"Authorization": "Token {0}".format(config.TOKEN)},  files={'file': file_obj})
+                    response.raise_for_status()
+                    self.uploaded_files += [f['filename']]
+            # Attempt to create node
+            self.add_nodes(node[0], node[1])
+
+    def check_failed(self):
```
Comment (jayoshih) [inline ricecooker/managers.py]: add docstring

File: ricecooker/utils/html.py [hunk 6]
```diff
@@ -0,0 +1,39 @@
+import os
+import requests
+import time
+
+from bs4 import BeautifulSoup
+
+from selenium import webdriver
+
+PHANTOMJS_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)), "..", "..", "node_modules", "phantomjs-prebuilt", "bin", "phantomjs")
```
Comment (jayoshih) [inline ricecooker/utils/html.py]: Do we require PhantomJS and Node now? Or only for HTML5 apps?

**Pull request merged**